api/v1beta1/cloudstackmachine_webhook.go

```diff
@@ -63,8 +63,8 @@ func (r *CloudStackMachine) ValidateCreate() error {
 		errorList = append(errorList, field.Forbidden(field.NewPath("spec", "identityRef", "kind"), "must be a Secret"))
 	}
 
-	errorList = webhookutil.EnsureFieldExists(r.Spec.Offering, "Offering", errorList)
-	errorList = webhookutil.EnsureFieldExists(r.Spec.Template, "Template", errorList)
+	errorList = webhookutil.EnsureAtLeastOneFieldExists(r.Spec.Offering.ID, r.Spec.Offering.Name, "Offering", errorList)
```

Comment:
I imagine that it'd be incorrect to specify both ID and Name.  Should this test be EnsureExactlyOneFieldExists?

Reply:
In this case, ID takes priority. If both ID and name are both provided, name will be ignored.

We have test cases to cover this. I had a brief discussion with Joshua about similar situation in multiple-zone (both zone id and name provided) and decided that ID will be used in this situation. So offering and template follow the same pattern.

Reply:
I'm not a fan of silently doing something different than what the user requested.  As a user I'd rather be told I'm doing something wrong, rather than pull my hair out trying to figure out why it was behaving differently than I expected.

Reply:
After thinking it over again, we can add an extra logic when both ID and name provided: 
1. we still use ID, 
2. we check the name provided with name returned from API call by using ID
3. we emit a warning if mismatch, but still continue the flow.